provider/mesh/state.go

```diff
@@ -0,0 +1,327 @@
+}
+
+// Merge the notification set with gossip data and return a new notification
+// state. The original state remains unchanged.
+// The state is based in LWW manner using the timestamp.
+func (st *notificationState) Merge(other mesh.GossipData) mesh.GossipData {
+	o := other.(*notificationState)
+	o.mtx.RLock()
+	defer o.mtx.RUnlock()
+
+	return st.mergeComplete(o.set)
+}
+
+func (st *notificationState) mergeComplete(set map[string]notificationEntry) *notificationState {
+	st.mtx.Lock()
+	defer st.mtx.Unlock()
+
+	for k, v := range set {
+		if prev, ok := st.set[k]; !ok || prev.Timestamp.Before(v.Timestamp) {
+			st.set[k] = v
+		}
+	}
+	// XXX(fabxc): from what I understand we merge into the receiver and what
+	// we return should be exactly that.
+	// As all operations are locked, this should be fine.
+	return st
+}
+
+func (st *notificationState) mergeDelta(set map[string]notificationEntry) *notificationState {
+	st.mtx.Lock()
+	defer st.mtx.Unlock()
+
+	d := map[string]notificationEntry{}
+
+	for k, v := range set {
+		if prev, ok := st.set[k]; !ok || prev.Timestamp.Before(v.Timestamp) {
+			st.set[k] = v
+			d[k] = v
+		}
+	}
+	return &notificationState{set: d}
+}
+
+type silenceState struct {
+	mtx   sync.RWMutex
+	m     map[uuid.UUID]*types.Silence
+	stopc chan struct{}
+	now   func() time.Time // now function for test injection
+}
+
+func newSilenceState() *silenceState {
+	return &silenceState{
+		m:     map[uuid.UUID]*types.Silence{},
+		stopc: make(chan struct{}),
+		now:   time.Now,
+	}
+}
+
+func (s *silenceState) run(retention time.Duration) {
+	for {
+		select {
+		case <-s.stopc:
+			return
+		case <-time.After(gcInterval):
+			s.gc(retention)
+		}
+	}
+}
+
+func (s *silenceState) stop() {
+	close(s.stopc)
+}
+
+func (s *silenceState) gc(retention time.Duration) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	t := s.now().Add(-retention)
+	for k, v := range s.m {
+		if v.EndsAt.Before(t) {
+			delete(s.m, k)
+		}
+	}
+}
+
+func decodeSilenceSet(b []byte) (map[uuid.UUID]*types.Silence, error) {
+	var v map[uuid.UUID]*types.Silence
+	err := gob.NewDecoder(bytes.NewReader(b)).Decode(&v)
+	return v, err
+}
+
+func (st *silenceState) Encode() [][]byte {
+	st.mtx.RLock()
+	defer st.mtx.RUnlock()
+
+	var buf bytes.Buffer
+	if err := gob.NewEncoder(&buf).Encode(&st.m); err != nil {
+		panic(err)
+	}
+	return [][]byte{buf.Bytes()}
+}
+
+// silenceModAllowed checks whether silence a may be changed to silence b.
+// Returns an error stating the reason if not.
+// The silences are guaranteed to be valid. Silence a may be nil if b is a new.
+func silenceModAllowed(a, b *types.Silence, now time.Time) error {
+	if a == nil {
+		if b.StartsAt.Before(now) {
+			// From b being valid it follows that EndsAt will also be
+			// in the future.
+			return fmt.Errorf("new silence may not start in the past")
+		}
+		return nil
+	}
+	if a.ID != b.ID {
+		return fmt.Errorf("IDs do not match")
+	}
+	if !b.StartsAt.Equal(a.StartsAt) {
```

Comment:
This runs into a precision problem (start time parsed from the UI will almost be slightly different from the start time in the silence itsilf).